### ISSUE TITLE
Discussion: Dependency injection vs globals

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3564]
## 🧰 
Something I think about a lot, but haven't written up anywhere (and would like to gather thoughts on) is the general usage pattern of global objects in napari.  Current global objects in napari include are "service"-like application related things such as:
- `napari.plugins.plugin_manager`
- `napari.utils.action_manager.action_manager`, 
- the settings object return by `napari.settings.get_settings`
- the global Qt application (and a number of other Qt-related things)

and as I play with patterns in `npe2`, it's clear that we'll either need more of these things, (e.g. a "keybinding" service probably should live independent of the "action/command" service, "context", "themes", and "menus" are also things that could nicely fit into a service-like interface), _or_ we'll start to make behemoth god-objects.

**Global objects** are obviously an easy target for criticism.  If one is not very careful, they can create bugs, such the button-stealing behavior we saw in #2897, #2977, and related discussions around #2914.  (A "pro" of globals is that they are extremely easy to reason about ... more on that in a minute).  I would argue that that particular case isn't strictly due to the evil of globals, but rather a leaking of button-instance-specific logic into the global object (and the very real difficulty of integrating something like the action manager into our already-existing keybinding framework without ripping it out and starting over).

One alternative to global objects is **dependency injection**: which, simply put, just means that rather than objects fetching/creating the things they depend on (e.g. the `QtViewerButtons` goes and gets the global `action_manager`), something _provides_ ("injects") that dependency into the object via its constructor or otherwise (e.g. `QtViewerButtons(self.viewer, self.action_manager)`).  A basic example of dependency injection is what @Czaki did in https://github.com/napari/napari/pull/3437, where you'll see an `ActionManager` instance being added as a parameter to a lot of things.

Dependency injection is a super-common pattern in large-scale applications, and it _does_ solve a lot of problems related to global objects (vscode, for example, uses services and dependency-injection liberally).  It can make unit testing a _lot_ easier, since you can break up your application into smaller interfaces, and mock those interfaces and inject those services into each individual component while testing (that is, you don't need to monkey patch a global object, you just pass in the dependency directly).

## `dependency_injector`
For napari as an "application", dependency injection seems to make a lot of sense to me.  But it does create challenges for using napari as a library: for example, we obviously don't want an end-user to have inject their own dependencies (like, create an `ActionManager()` and pass it to a `Viewer()` or something).  It can also make it a bit harder to reason about, since it may be less clear where the specific dependency/instance is coming from.

There are frameworks out there to make this stuff easier, and this is part of what I want to get feedback on.
See `dependency_injector`: https://python-dependency-injector.ets-labs.org/index.html

From my initial reading, the idea there is that you have a **`Container` that declares all of the services/dependencies/"providers" that your application has**:

_some syntax is probably wrong here... just an idea example_
```py
from dependency_injector import containers, providers

class Container(containers.DeclarativeContainer):
    settings = providers.Configuration()
    plugin_manager = providers.Singleton(PluginManager)
    # dependencies can depend on other services
    action_manager = providers.Factory(ActionManager, plugin_manager=plugin_manager)
```

then **objects use type hints to declare their dependencies**:

```py
from dependency_injector.wiring import Provide

class ActionManager:
    def __init__(self, plugin_manager: PluginManager = Provide[Container.plugin_manager]): ...
```

and finally, **the whole application is run in the _context_ of some container**:

```py
from dependency_injector.wiring import inject, Provide

@inject
def main(plugin_manager: PluginManager = Provide[Container.plugin_manager]):
    ...

if __name__ == '__main__':
    container = Container()
    main()  # <-- dependency is injected automatically

    # or, specific things can be overridden / mocked
    with container.api_client.override(mock.Mock()):
        main()  # <-- overridden dependency is injected automatically
```

Note: a pre-requisite for me would be that no one ever has to create a dumb dependency just to instantiate something in the absence of the full application... so it would have to be possible to create a one-off dependency (not immediately sure how the framework would do that)

you can see a more complete application example [here](https://python-dependency-injector.ets-labs.org/examples/application-single-container.html)

curious to get thoughts.  @jni, @sofroniewn, @goanpeca, @Carreau, @nclack, @alisterburt, @andy-sweet, @Czaki

**Comment by andy-sweet** [comment]
This is an excellent summary of DI (dependency injection) and some of the pros/cons of it as a pattern, and some benefits of using DI frameworks.

I recently developed on a project using a DI framework ([Dagger](https://dagger.dev/) in Java/Android), so I'm happy to share some of my experiences.

First, it's important to clarify/reiterate the difference between DI and DI frameworks. DI is mostly about separating behavior and dependency resolution, which encourages more modular and composable design. When using DI, one that class that depends on another does not construct/get an instance of it - instead preparing the instance of the dependency is done elsewhere, and the instance (or a provider for it) is passed through (usually in the constructor).

One of the main motivations for DI frameworks is that the DI pattern usually creates a lot of boilerplate code because something somewhere has to create the instances of dependencies that are needed. In some cases, this boilerplate code can hurt the readability of the main logic of a program. DI frameworks define dedicated places to define how dependencies should be constructed/resolved (like the `Container` above), which also allows you to adapt this based on the context (e.g. production vs. test environment). Another potential benefit is that frameworks will often allow you to configure lazy construction or will do it by default.

For those interested, I think [the Guice User Guide's Motivation document](https://github.com/google/guice/wiki/Motivation) is pretty good at explaining why you would want to use DI and a DI framework in general, though you'll have to stomach some Java to read through the examples.

The biggest downside of a DI framework in my experience is that it adds a learning curve for developers on a project, if they're not familiar with DI or the specific framework being used. I think that's particularly relevant for new contributors on an open source science software project. As an aside, this is a common problem when replacing boilerplate code that isn't mentioned enough - it's only boilerplate code to those that have seen it a lot before. Some ways to mitigate this learning curve are:
- Only pick a framework that is well documented with examples (`dependency_injector` looks pretty good from a first look).
- Be prepared to write our own napari guide on using the DI framework that we use.
- Limit use of DI to the implementation guts of napari and hide it from the main public API (this is probably pretty hard because once you start use a DI framework it tends to spread over all the code base).
- Restrict our usage of a particular framework. Some frameworks (e.g. Dagger) allow you to configure a lot (e.g. custom scopes) that can make reasoning about construction difficult. Using/Encouraging a simpler subset of the functionality of a framework could help.

**Comment by Czaki** [comment]
I do not have experience using any dependecy incetor framework but it looks promising. But we still need remember about nature of Qt and event base programming. 

Is someone will create napari Viewer in application and then execute some other code (like PartSeg > View > Napari viewer) and there is a need of pass context calculated after start of Qt event loop (or for example start napari in `ipython --gui=qt`) then modified context will be available only during create of viewer and will not affect functions called from events. So objects still should remember reference to bocject during creation and update context on function call. 

Because of this I'm not sure if direct passing of needed objects does not needs less changes of code. But There are examples, like creating of preference window from #3437 where context manager may be a nice thing?

**Comment by tlambert03** [comment]
Thanks for that great additional detail @andy-sweet!

> First, it's important to clarify/reiterate the difference between DI and DI frameworks.  

very good point.  I also wanted to mention that many libraries have their own internal DI patterns that don't necessarily require an external framework.  (FastAPI, for example, has a [nice internal DI pattern](https://fastapi.tiangolo.com/tutorial/dependencies/))...  I was playing with it today and got something working pretty well with:

```py
class A:
    ...

@inject
class B:
    def __init__(self, a: A = Depends()): 
        self.a = a

b = B()
assert isinstance(b.a, A)
```

so it seems conceivable that we could start small and restrictive?


> Limit use of DI to the implementation guts of napari and hide it from the main public API 

totally, agree, and not sure it would actually be all that hard.  From posts of forums, i really think that `Viewer` and `Layer` methods make up the vast majority of our API consumption, and those remain relatively nicely "model-based", and pretty clean of the application-focused things that really need the dependencies.  

I think a clarifying step here would probably be to actually make something of a dependency tree, to clarify what actually depends on what, and perhaps even try to nail down and minimize those interfaces.

**Comment by Czaki** [comment]
As far as I understand the use of a dependency injector is suggested as a solution for using napari as a library, not for plugin creator. The core of the problem with the global objects is using multiple viewer instances and interactions between them through global objects?

**Comment by tlambert03** [comment]
> The core of the problem with the global objects is using multiple viewer instances and interactions between them through global objects?

I still just don't know that I entirely agree that this is _the core_ problem. I definitely see that this was _a_ problem, specifically with regards to the action manager and buttons, but, as mentioned in my first post I really do see that as a specific difficulty with the design of the current action manager, that "took on too much" by managing literal button instances in individual viewers.  Take that specific problem away and I think there's a more interesting, less obvious, and broader discussion to be had

**Comment by tlambert03** [comment]
> As far as I understand the use of a dependency injector is suggested as a solution for using napari as a library, not for plugin creator.

yep, definitely not aiming external users.  this is simply a question of how best to organize the rapidly growing number of interfaces and objects that napari manages

**Comment by Czaki** [comment]
Looking in the more general case. As far as I understand dependency injector is the solution for temporary change behavior of some part of code? And from my point of view, in most of the cases, I would like to have the possibility to change behavior for instance, mot time unit. 

Also, did you test if the suggested library plays nice with pydantic? I observe that `monkeypatch` from `pytest` fails to replace wield of a pydantic model.
